### PR TITLE
Nullable in comparers

### PR DESCRIPTION
**Pull request by SimonCropp**

### CODE REVIEW

File: src/NUnitFramework/framework/Constraints/Comparers/KeyValuePairsComparer.cs [hunk 1]
```diff
@@ -18,21 +19,26 @@ internal static class KeyValuePairsComparer
...
+            object? yKey = yType.GetProperty("Key")!.GetValue(y, null);
 
-            return equalityComparer.AreEqual(xKey, yKey, ref keyTolerance, state.PushComparison(x, y)) 
-                && equalityComparer.AreEqual(xValue, yValue, ref tolerance, state.PushComparison(x, y));
+            if (!equalityComparer.AreEqual(xKey, yKey, ref keyTolerance, state.PushComparison(x, y)))
+            {
+                return false;
+            }
+
+            object? xValue = xType.GetProperty("Value")!.GetValue(x, null);
```
Comment (SimonCropp) [inline src/NUnitFramework/framework/Constraints/Comparers/KeyValuePairsComparer.cs:38]: minor perf tweak that avoids some reflection is the keys are not equal

File: src/NUnitFramework/framework/Constraints/Comparers/StringsComparer.cs [hunk 2]
```diff
@@ -13,7 +13,7 @@ internal static class StringsComparer
     {
         public static bool? Equal(object x, object y, ref Tolerance tolerance, ComparisonState state, NUnitEqualityComparer equalityComparer)
         {
-            if (!(x is string xString) || !(y is string yString))
+            if (x is not string xString || y is not string yString)
```
Comment (stevenaw) [inline src/NUnitFramework/framework/Constraints/Comparers/StringsComparer.cs:16]: 👍 Thanks for cleaning all these up in the comparers too

**Review by stevenaw — APPROVED** [review]
Thank you once again @SimonCropp

**Comment by SimonCropp** [comment]
can we get a merge?

File: src/NUnitFramework/framework/Constraints/Comparers/ComparisonState.cs [hunk 3]
```diff
@@ -9,7 +10,7 @@ namespace NUnit.Framework.Constraints.Comparers
         /// <summary>
         /// Flag indicating whether or not this is the top level comparison.
         /// </summary>
-        public readonly bool TopLevelComparison { get; }
+        public bool TopLevelComparison { get; }
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Constraints/Comparers/ComparisonState.cs:13]: Just to be sure - I haven't worked much with `readonly ref structs`, `readonly structs`, nor `ref structs` - these are removed as they are redundant due to the `readonly` on the class?

Comment (SimonCropp) [reply src/NUnitFramework/framework/Constraints/Comparers/ComparisonState.cs:13]: yep. that’s correct

**Review by mikkelbu — APPROVED** [review]

**Pull request merged**